includes/core/classes/endpoints/class-endpoint.php

```diff
@@ -0,0 +1,513 @@
+	 *
+	 * @return bool True if the query is valid, false otherwise.
+	 */
+	public function is_valid_query(): bool {
+		return ( $this->validation_callback )() && ! empty( get_query_var( $this->query_var ) );
+	}
+
+	/**
+	 * Retrieves the slugs of the specified endpoint types.
+	 *
+	 * This method filters the `types` array to get the slugs for either a specific type of endpoint
+	 * (e.g., `Endpoint_Redirect` or `Endpoint_Template`) or returns slugs for all types if no type
+	 * is specified.
+	 *
+	 * @since 1.0.0
+	 *
+	 * @param string|null $entity Optional. The class name of the endpoint type to filter by (e.g., 'Endpoint_Redirect' or 'Endpoint_Template').
+	 *                            If null, it retrieves slugs for all types.
+	 * @return string[]           An array of slugs for the specified or all types.
+	 */
+	protected function get_slugs( ?string $entity = null ): array {
+		// Determine Endpoint_Types to get slug names from.
+		$types = ( null === $entity )
+			// All?
+			? $this->types
+			// Or a specific type?
+			: array_filter(
+				$this->types,
+				function ( $type ) use ( $entity ) {
+					return self::is_of_class( $type, $entity );
+				}
+			);
+		return wp_list_pluck( $types, 'slug' );
+	}
+
+	/**
+	 * Checks if the given endpoint type is an instance of the specified class.
+	 *
+	 * This method verifies whether the provided `$type` is an instance of the `$entity`
+	 * class. It first checks if the `$entity` exists in the defined list of valid endpoint
+	 * classes by calling `is_in_class()`. If the entity is valid, it further checks if the
+	 * `$type` is an instance of that class.
+	 *
+	 * @since 1.0.0
+	 *
+	 * @param  Endpoint_Type $type   The endpoint type object to check.
+	 * @param  string        $entity The class name of the entity to check against (e.g., 'Endpoint_Redirect' or 'Endpoint_Template').
+	 * @return bool                  True if the `$type` is an instance of the `$entity` class, false otherwise.
+	 */
+	private static function is_of_class( Endpoint_Type $type, string $entity ): bool {
+		return self::is_in_class( $entity ) && $type instanceof $entity;
+	}
+
+	/**
+	 * Checks if the given entity is a valid endpoint class in the current namespace.
+	 *
+	 * This method verifies whether the provided `$entity` exists in the predefined list
+	 * of valid endpoint classes within the current namespace. It helps ensure that only
+	 * valid classes (like `Endpoint_Redirect` or `Endpoint_Template`) are used when
+	 * checking endpoint types.
+	 *
+	 * @since 1.0.0
+	 *
+	 * @param  string $entity The class name of the entity to check (e.g., 'Endpoint_Redirect' or 'Endpoint_Template').
+	 * @return bool           True if the `$entity` is a valid endpoint class, false otherwise.
+	 */
+	private static function is_in_class( string $entity ): bool {
+		return in_array(
+			$entity,
+			array(
+				__NAMESPACE__ . '\Endpoint_Redirect',
+				__NAMESPACE__ . '\Endpoint_Template',
+			),
+			true
+		);
+	}
+}
```

Comment:
This two methods also look a lot like as they were better placed in a more common place.